velox/dwio/common/BitPackDecoderUtil.h

```diff
@@ -0,0 +1,600 @@
+
+#include "velox/vector/TypeAliases.h"
+
+#include <folly/Range.h>
+#include <xsimd/config/xsimd_config.hpp> // @manual
+
+namespace facebook::velox::dwio::common {
+
+static constexpr const uint64_t kPdepMask8[] = {
+    0x0000000000000000,
+    0x0101010101010101,
+    0x0303030303030303,
+    0x0707070707070707,
+    0x0f0f0f0f0f0f0f0f,
+    0x1f1f1f1f1f1f1f1f,
+    0x3f3f3f3f3f3f3f3f,
+    0x7f7f7f7f7f7f7f7f,
+    0xffffffffffffffff};
+
+// Pdep instruction masks for uint16_t. Generated by generatePdepMasks(16).
+static constexpr const uint64_t kPdepMask16[] = {
+    0x0000000000000000,
+    0x0001000100010001,
+    0x0003000300030003,
+    0x0007000700070007,
+    0x000f000f000f000f,
+    0x001f001f001f001f,
+    0x003f003f003f003f,
+    0x007f007f007f007f,
+    0x00ff00ff00ff00ff,
+    0x01ff01ff01ff01ff,
+    0x03ff03ff03ff03ff,
+    0x07ff07ff07ff07ff,
+    0x0fff0fff0fff0fff,
+    0x1fff1fff1fff1fff,
+    0x3fff3fff3fff3fff,
+    0x7fff7fff7fff7fff,
+    0xffffffffffffffff};
+
+// Pdep instruction masks for uint32_t. Generated by generatePdepMasks(32).
+static constexpr const uint64_t kPdepMask32[] = {
+    0x0000000000000000, 0x0000000100000001, 0x0000000300000003,
+    0x0000000700000007, 0x0000000f0000000f, 0x0000001f0000001f,
+    0x0000003f0000003f, 0x0000007f0000007f, 0x000000ff000000ff,
+    0x000001ff000001ff, 0x000003ff000003ff, 0x000007ff000007ff,
+    0x00000fff00000fff, 0x00001fff00001fff, 0x00003fff00003fff,
+    0x00007fff00007fff, 0x0000ffff0000ffff, 0x0001ffff0001ffff,
+    0x0003ffff0003ffff, 0x0007ffff0007ffff, 0x000fffff000fffff,
+    0x001fffff001fffff, 0x003fffff003fffff, 0x007fffff007fffff,
+    0x00ffffff00ffffff, 0x01ffffff01ffffff, 0x03ffffff03ffffff,
+    0x07ffffff07ffffff, 0x0fffffff0fffffff, 0x1fffffff1fffffff,
+    0x3fffffff3fffffff, 0x7fffffff7fffffff, 0xffffffffffffffff};
+
+static const uint32_t BITPACK_MASKS[] = {
+    0,          1,          3,         7,         15,        31,
+    63,         127,        255,       511,       1023,      2047,
+    4095,       8191,       16383,     32767,     65535,     131071,
+    262143,     524287,     1048575,   2097151,   4194303,   8388607,
+    16777215,   33554431,   67108863,  134217727, 268435455, 536870911,
+    1073741823, 2147483647, 4294967295};
+
+/// Unpack numValues number of input values from inputBuffer. The results
+/// will be written to result. numValues must be a multiple of 8. The
+/// caller needs to make sure the inputBufferLen contains at least numValues
+/// number of packed values. The inputBits and result pointers will be updated
+/// to the next to read position after this call.
+template <typename T>
+static inline uint32_t unpackNaive(
+    const uint8_t* FOLLY_NONNULL& inputBits,
+    uint64_t inputBufferLen,
+    uint64_t numValues,
+    uint8_t bitWidth,
+    T* FOLLY_NONNULL& result);
+
+// The function definitions are put here to make sure they are inlined. Moving
+// them to the .cpp file may result in 10x regression.
+
```

Comment:
@Yuhta this comment is directly porting from previous AVX2 implements, it also indicated that moving the functions to .cpp file may result in 10x regression